File: services/integration/sync_transaction.js
/**
 * Sync Transaction Manager
 * 
 * Manages the lifecycle of synchronization transactions, including
 * tracking, commit/rollback, and persistence.
 */

const { logger } = require('../../shared/logger');
const { db } = require('../../shared/db');

class SyncTransaction {
  /**
   * Create a new sync transaction
   * 
   * @param {Object} event - The synchronization event
   */
  constructor(event) {
    this.event = event;
    this.transactionId = event.transactionId;
    this.entityType = event.entityType;
    this.operation = event.operation;
    this.sourceSystem = event.sourceSystem;
    this.targetSystem = event.targetSystem;
    this.startTime = new Date();
    this.status = 'pending';
    this.steps = [];
  }

  /**
   * Begin the transaction
   */
  async begin() {
    logger.info(`Beginning sync transaction: ${this.transactionId}`);
    
    // Record transaction start in database
    await db.syncTransactions.create({
      transactionId: this.transactionId,
      entityType: this.entityType,
      operation: this.operation,
      sourceSystem: this.sourceSystem,
      targetSystem: this.targetSystem,
      startTime: this.startTime,
      status: this.status,
      eventData: JSON.stringify(this.event)
    });
    
    return this;
  }

  /**
   * Record a step in the transaction
   * 
   * @param {string} description - Description of the step
   * @param {Object} data - Additional data for the step
   */
  async recordStep(description, data = {}) {
    const step = {
      timestamp: new Date(),
      description,
      data
    };
    
    this.steps.push(step);
    
    // Update transaction in database
    await db.syncTransactionSteps.create({
      transactionId: this.transactionId,
      timestamp: step.timestamp,
      description: step.description,
      stepData: JSON.stringify(step.data)
    });
    
    return this;
  }

  /**
   * Commit the transaction
   */
  async commit() {
    logger.info(`Committing sync transaction: ${this.transactionId}`);
    this.status = 'completed';
    
    // Update transaction in database
    await db.syncTransactions.update(
      { 
        status: this.status,
        endTime: new Date(),
        duration: new Date() - this.startTime
      },
      { where: { transactionId: this.transactionId } }
    );
    
    return true;
  }

  /**
   * Roll back the transaction
   * 
   * @param {Error} error - Error that caused the rollback
   */
  async rollback(error) {
    logger.error(`Rolling back sync transaction: ${this.transactionId}. Error: ${error.message}`);
    this.status = 'failed';
    
    // Update transaction in database
    await db.syncTransactions.update(
      { 
        status: this.status,
        endTime: new Date(),
        duration: new Date() - this.startTime,
        errorMessage: error.message
      },
      { where: { transactionId: this.transactionId } }
    );
    
    // Implement any necessary cleanup or compensating actions
    // This would be specific to each entity type and operation
    
    return true;
  }

  /**
   * Get transaction by ID
   * 
   * @param {string} transactionId - The transaction ID to retrieve
   * @returns {Object} - The transaction data
   */
  static async getById(transactionId) {
    const transaction = await db.syncTransactions.findOne({
      where: { transactionId },
      include: [{
        model: db.syncTransactionSteps,
        as: 'steps'
      }]
    });
    
    return transaction;
  }

  /**
   * List recent transactions
   * 
   * @param {Object} filters - Optional filters for the query
   * @param {number} limit - Max number of transactions to return
   * @returns {Array} - List of transactions
   */
  static async listRecent(filters = {}, limit = 100) {
    const where = {};
    
    if (filters.status) {
      where.status = filters.status;
    }
    
    if (filters.entityType) {
      where.entityType = filters.entityType;
    }
    
    if (filters.sourceSystem) {
      where.sourceSystem = filters.sourceSystem;
    }
    
    const transactions = await db.syncTransactions.findAll({
      where,
      order: [['startTime', 'DESC']],
      limit,
      include: [{
        model: db.syncTransactionSteps,
        as: 'steps'
      }]
    });
    
    return transactions;
  }
}

module.exports = {
  SyncTransaction
};

File: services/integration/sync_transaction.rs
// services/integration/sync_transaction.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use std::error::Error;
use uuid::Uuid;

use crate::shared::logger::Logger;
use crate::shared::db::Database;

/// Sync Transaction Manager
///
/// Manages the lifecycle of synchronization transactions, including
/// tracking, commit/rollback, and persistence.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SyncTransaction {
    pub transaction_id: String,
    pub entity_type: String,
    pub operation: String,
    pub source_system: String,
    pub target_system: String,
    pub start_time: DateTime<Utc>,
    pub end_time: Option<DateTime<Utc>>,
    pub status: String,
    pub steps: Vec<TransactionStep>,
    pub event: serde_json::Value,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TransactionStep {
    pub step_id: String,
    pub name: String,
    pub status: String,
    pub timestamp: DateTime<Utc>,
    pub error: Option<String>,
    pub data: Option<serde_json::Value>,
}

impl SyncTransaction {
    /// Create a new sync transaction
    pub fn new(event: serde_json::Value) -> Self {
        let transaction_id = event["transactionId"].as_str()
            .unwrap_or_else(|| Uuid::new_v4().to_string().as_str())
            .to_string();
            
        let entity_type = event["entityType"].as_str()
            .unwrap_or("unknown")
            .to_string();
            
        let operation = event["operation"].as_str()
            .unwrap_or("unknown")
            .to_string();
            
        let source_system = event["sourceSystem"].as_str()
            .unwrap_or("unknown")
            .to_string();
            
        let target_system = event["targetSystem"].as_str()
            .unwrap_or("unknown")
            .to_string();
        
        SyncTransaction {
            transaction_id,
            entity_type,
            operation,
            source_system,
            target_system,
            start_time: Utc::now(),
            end_time: None,
            status: "pending".to_string(),
            steps: Vec::new(),
            event,
        }
    }
    
    /// Begin the transaction
    pub async fn begin(&mut self, db: &mut Database) -> Result<(), Box<dyn Error>> {
        let logger = Logger::new("SyncTransaction");
        logger.info(&format!("Beginning sync transaction: {}", self.transaction_id), None);
        
        // Record transaction start in database
        db.sync_transactions.create(serde_json::json!({
            "transactionId": self.transaction_id,
            "entityType": self.entity_type,
            "operation": self.operation,
            "sourceSystem": self.source_system,
            "targetSystem": self.target_system,
            "startTime": self.start_time,
            "status": self.status,
        })).await?;
        
        Ok(())
    }
    
    /// Add a step to the transaction
    pub fn add_step(&mut self, name: &str, status: &str, data: Option<serde_json::Value>) -> TransactionStep {
        let step = TransactionStep {
            step_id: Uuid::new_v4().to_string(),
            name: name.to_string(),
            status: status.to_string(),
            timestamp: Utc::now(),
            error: None,
            data,
        };
        
        self.steps.push(step.clone());
        step
    }
    
    /// Mark a step as complete
    pub fn complete_step(&mut self, step_id: &str) -> Option<&mut TransactionStep> {
        if let Some(step) = self.steps.iter_mut().find(|s| s.step_id == step_id) {
            step.status = "completed".to_string();
            Some(step)
        } else {
            None
        }
    }
    
    /// Mark a step as failed
    pub fn fail_step(&mut self, step_id: &str, error: &str) -> Option<&mut TransactionStep> {
        if let Some(step) = self.steps.iter_mut().find(|s| s.step_id == step_id) {
            step.status = "failed".to_string();
            step.error = Some(error.to_string());
            Some(step)
        } else {
            None
        }
    }
    
    /// Commit the transaction
    pub async fn commit(&mut self, db: &mut Database) -> Result<(), Box<dyn Error>> {
        let logger = Logger::new("SyncTransaction");
        logger.info(&format!("Committing sync transaction: {}", self.transaction_id), None);
        
        self.status = "committed".to_string();
        self.end_time = Some(Utc::now());
        
        // Update transaction in database
        db.sync_transactions.update(
            self.transaction_id.clone(),
            serde_json::json!({
                "status": self.status,
                "endTime": self.end_time,
                "steps": self.steps,
            })
        ).await?;
        
        Ok(())
    }
    
    /// Rollback the transaction
    pub async fn rollback(&mut self, db: &mut Database, reason: &str) -> Result<(), Box<dyn Error>> {
        let logger = Logger::new("SyncTransaction");
        logger.warn(&format!("Rolling back sync transaction: {} - Reason: {}", 
            self.transaction_id, reason), None);
        
        self.status = "rolled_back".to_string();
        self.end_time = Some(Utc::now());
        
        // Add rollback step
        self.add_step("rollback", "completed", Some(serde_json::json!({
            "reason": reason
        })));
        
        // Update transaction in database
        db.sync_transactions.update(
            self.transaction_id.clone(),
            serde_json::json!({
                "status": self.status,
                "endTime": self.end_time,
                "steps": self.steps,
            })
        ).await?;
        
        Ok(())
    }
}
